src/wasmtime.rs

```diff
@@ -100,6 +114,91 @@ fn read_wasm(path: PathBuf) -> Result<Vec<u8>, String> {
     })
 }
 
+fn compute_preopen_dirs(flag_dir: &[String], flag_mapdir: &[String]) -> Vec<(String, libc::c_int)> {
+    let mut preopen_dirs = Vec::new();
+
+    for dir in flag_dir {
+        let fd = unsafe {
+            libc::open(
+                CString::new(dir.as_bytes()).unwrap().as_ptr(),
+                libc::O_RDONLY | libc::O_DIRECTORY,
+            )
+        };
+        if fd < 0 {
+            println!("error while pre-opening directory {}: {}", dir, errno());
+            exit(1);
+        }
+
+        preopen_dirs.push((dir.clone(), fd));
+    }
+
+    for mapdir in flag_mapdir {
+        let parts: Vec<&str> = mapdir.split(':').collect();
+        if parts.len() != 2 {
+            println!("--mapdir argument must contain exactly one colon, separating a guest directory name and a host directory name");
+            exit(1);
+        }
+        let (key, value) = (parts[0], parts[1]);
+        let fd = unsafe {
+            libc::open(
+                CString::new(value.as_bytes()).unwrap().as_ptr(),
+                libc::O_RDONLY | libc::O_DIRECTORY,
+            )
+        };
+        if fd < 0 {
+            println!("error while pre-opening directory {}: {}", value, errno());
+            exit(1);
+        }
+
+        preopen_dirs.push((key.to_string(), fd));
+    }
+
+    preopen_dirs
+}
+
+/// Compute the argv array values.
+fn compute_argv(argv0: &str, arg_arg: &[String]) -> Vec<String> {
+    let mut result = Vec::new();
+
+    // Add argv[0], which is the program name. Only include the base name of the
+    // main wasm module, to avoid leaking path information.
+    result.push(
+        Path::new(argv0)
+            .components()
+            .next_back()
+            .map(Component::as_os_str)
```

Comment:
You can use `path.file_name()`: https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.file_name

Reply:
Ah, I was looking for a function that did that, but somehow missed it. Thanks!